src/go/rpk/pkg/cli/cmd/root.go

```diff
@@ -89,3 +104,80 @@ func Execute() {
 		os.Exit(1)
 	}
 }
+
+type pluginHandler interface {
+	lookPath(file string) (path string, ok bool)
+	exec(path string, args []string) error
+}
+
+// tryExecPlugin looks for a plugin, following the exact behavior of kubernetes:
+//
+//  - all dashes are turned to underscores
+//  - all "pieces" (non-flags) are joined with a dash
+//  - we prefer the longest command match
+//  - we search upward by piece until we run out of pieces
+//  - lastly, the command must be executable
+//
+// Dashes are only replaced into underscores during the path search; if an
+// argument hash a dash and is used as an argument to a plugin, the dash
+// remains.
+//
+// So,
+//
+//     rpk foo-bar baz boz fizz-buzz --flag
+//
+// is translated into searching and execing (with osPluginHandler), in order:
+//
+//     rpk-foo_bar-baz-boz-fizz_buzz (with args "--flag")
+//     rpk-foo_bar-baz-boz           (with args "fizz-buzz --flag")
+//     rpk-foo_bar-baz               (with args "boz fizz-buzz --flag")
+//     rpk-foo_bar                   (with args "baz boz fizz-buzz --flag")
+//
+// If a plugin is run, this returns the run error and true, otherwise this
+// returns false.
+func tryExecPlugin(h pluginHandler, args []string) (string, error) {
+	var pieces []string
+	for _, arg := range args {
+		if strings.HasPrefix(arg, "-") { // found a flag, quit
+			break
+		}
+		pieces = append(pieces, strings.Replace(arg, "-", "_", -1))
+	}
+
+	if len(pieces) == 0 {
+		return "", nil // no plugin specified (command is just "rpk")
+	}
+
+	foundPath := ""
+	for len(pieces) > 0 {
+		path, ok := h.lookPath("rpk-" + strings.Join(pieces, "-"))
+		if !ok { // did not find with this piece, strip and search higher
+			pieces = pieces[:len(pieces)-1]
+			continue
+		}
+		foundPath = path
+		break
+	}
+	if len(foundPath) == 0 {
+		return "", nil
+	}
+
+	return foundPath, h.exec(foundPath, args[len(pieces):])
+}
+
+type osPluginHandler struct{}
+
+func (*osPluginHandler) lookPath(file string) (string, bool) {
+	path, err := exec.LookPath(file)
+	return path, err == nil
+}
+func (*osPluginHandler) exec(path string, args []string) error {
+	return (&exec.Cmd{
```

Comment:
don't we want to override the process image with `execlpe()` or `execvpe()` or smth like that, instead of running a command. What's the point in bubbling the error instead of overriding the image.

Reply:
kubernetes itself does `syscall.Exec` on non-windows systems, but apparently windows does not support `syscall.Exec`: https://github.com/kubernetes/kubernetes/blob/12e3860cb084ef99f79eb2bc7c123f9521342aff/staging/src/k8s.io/kubectl/pkg/cmd/cmd.go#L408-L427

What does execlpe or execvpe buy over `exec.Cmd` here? AFAICT in the Go source this is doing a fork&exec, and I figured it may be nice to see that `rpk` executed a command when looking at pstree.

Also w.r.t. bubbling the error, that's just to get a nice error at the top, if the fork&exec fails. An alternative would be not bubbling, or not printing the error.

Reply:
the difference is `execlpe/vpe` will *override* the process image, so it *becomes* the child process. I think syscall.Exec is what i would expect as well. There is no child process in the exec mode. 

I think what Kubernetes does is the right thing.

Reply:
Right that makes sense, but what's the benefit of a replacement vs a child? Just trying to understand the motivation; switching to syscall.Exec is easy enough.

Reply:
child can have zombie processes if not dealt with correctly. Signal handlers not propagated to child correctly (kill -9) of the parent will not kill the child procs and leave them as zombies.  things like that.

Reply:
once nice thing about not going nuclear with exec(3) is that I believe golang's exec package has support for command cancellation via the command context. with vanilla exec(3) we'd have no ability to enforce nice policies on plugin writers like not blocking forever.

Reply:
+1 on the `exec.CommandContext()`

Reply:
`CommandContext` is less useful here because a plugin is not meant to be scoped, there is no expiry / nothing to cancel on the context.

I'll add syscall.Exec the same way kubernetes does it, while having the windows exception.

Reply:
agreed w/ @twmb  - there is no supervision policies that make sense to me.